feat(api): Add support for new runner creation API

lgatellier opened the pull request:
## Changes

This PR adds support for GitLab's new runner creation API `POST /user/runners`.
Closes #2606 

### Documentation and testing

- [X] Documentation in the matching [docs section](https://github.com/python-gitlab/python-gitlab/tree/main/docs)
- [X] [Unit tests](https://github.com/python-gitlab/python-gitlab/tree/main/tests/unit)

lgatellier commented:
Not sure if unit tests fits into `test_runners.py`. Could also be into ` test_users.py`.

Do not hesitate to tell me if I should move them.

JohnVillalovos commented:
> Not sure if unit tests fits into `test_runners.py`. Could also be into `test_users.py`.
> 
> Please tell me if I should move them.

In my opinion it should be in the `test_users.py` file since that is where the code is that it is testing.

But otherwise this looks good to me.

lgatellier commented:
Hi @JohnVillalovos, and thanks for your feedback !

I moved the unit test into ` test_users.py` , and fixed a little doc issue in my project runner creation example (which used `group_id` instead of `project_id`).

JohnVillalovos commented on gitlab/client.py:
```diff
@@ -189,6 +189,8 @@ def __init__(
         """See :class:`~gitlab.v4.objects.PagesDomainManager`"""
         self.user_activities = objects.UserActivitiesManager(self)
         """See :class:`~gitlab.v4.objects.UserActivitiesManager`"""
+        self.user_runners = objects.UserRunnerManager(self)
```
So I don't think this should be added here. I think it should be done how `UserEmail` is done.

nejch replied:
See https://github.com/python-gitlab/python-gitlab/pull/2635#discussion_r1299316773 - @lgatellier I think John here meant `CurrentUserEmail` and other endpoints related to the authenticated user:

https://github.com/python-gitlab/python-gitlab/blob/d054917ccb3bbcc9973914409b9e34ba9301663a/gitlab/v4/objects/users.py#L128-L135

nejch replied:
@lgatellier I tend to agree with John here we like to group our manages and objects according to how GitLab structures its REST API endpoints.

But I understand this is frustrating because it becomes a 2-way step now (create the runner via the user endpoint, then lazily get the created runner for manipulation via the root endpoint). I'm also unsure why the endpoint is done like that upstream as it will be inconsistent with the other runner requests now :confused:

JohnVillalovos replied:
What @nejch said 🙂

lgatellier replied:
Thanks @nejch  @JohnVillalovos  for your feedbacks !

I applied your suggestions. :wink:

JohnVillalovos commented on gitlab/v4/objects/users.py:
```diff
@@ -678,3 +680,28 @@ class UserFollowingManager(ListMixin, RESTManager):
     _path = "/users/{user_id}/following"
     _obj_cls = User
     _from_parent_attrs = {"user_id": "id"}
+
```
In `class User()` should add an attribute `runners` that is of type `UserRunnerManager`

Then the call would be:  `user.runners.create(...)`

nejch replied:
@JohnVillalovos just to be sure: I think you meant `CurrentUser` (i.e. all endpoints relating to `/user/` not `/users/`).

JohnVillalovos replied:
Ah. I think you are correct. Thanks @nejch !

JohnVillalovos commented on tests/unit/objects/test_users.py:
```diff
@@ -304,3 +318,9 @@ def test_list_starred_projects(user, resp_starred_projects):
     assert projects[0].id == project_content["id"]
+
+
+def test_create_user_runner(gl: gitlab.Gitlab, resp_runner_create):
```
Probably the test should be something like.

```
runner = user.runners.create({"runner_type": "instance_type"})
```

JohnVillalovos commented on docs/gl_objects/runners.rst:
```diff
@@ -71,6 +71,14 @@ Register a new runner::
+   workflow comes with a new API endpoint to create runner, which does not use
+   registration tokens.
+
+   The new endpoint can be called using ``user_runners.create()``.
```
Change it to user.runners.create()

JohnVillalovos commented on docs/gl_objects/users.rst:
```diff
@@ -456,3 +456,60 @@ Get the users activities::
+
+Create an instance-wide runner::
+
+    runner = gl.user_runners.create({
```
Update all of these to be `user.runners.create`

JohnVillalovos commented:
@lgatellier Thanks for the PR!

Left  comment to have it work like the other endpoints that use `/user/*`. Please let me know if you disagree with my thoughts.

nejch commented on docs/gl_objects/runners.rst:
```diff
@@ -71,6 +71,14 @@ Register a new runner::
+.. note::
+
+   A new runner registration workflow has been introduced since GitLab 16.0. This new
+   workflow comes with a new API endpoint to create runner, which does not use
```
```suggestion
   workflow comes with a new API endpoint to create runners, which does not use
```

nejch commented on docs/gl_objects/runners.rst:
```diff
@@ -71,6 +71,14 @@ Register a new runner::
+   workflow comes with a new API endpoint to create runner, which does not use
+   registration tokens.
+
+   The new endpoint can be called using ``user.runners.create()``.
```
```suggestion
   The new endpoint can be called using ``gl.user.runners.create() after authenticating with `gl.auth()```.
```

nejch commented on tests/unit/objects/test_users.py:
```diff
@@ -241,6 +241,19 @@ def resp_starred_projects():
+            url="http://localhost/api/v4/user/runners",
+            json={"id": "6", "token": "6337ff461c94fd3fa32ba3b1ff4125"},
+            content_type="application/json",
+            status=200,
```
```suggestion
            status=201,
```

nejch commented:
Thanks a lot @lgatellier @JohnVillalovos I just pushed a minor typo suggestion so that we can get this in and continue with the tasks before the next scheduled release. If you see anything else just open a follow-up :)
